### DISCUSSION TITLE
MQTT verwenden...

### DISCUSSION THREAD

**Discussion by wolwes** [tbnobody/OpenDTU#750] [Q&A]
ich beschäftige mich schon seit geraumer Zeit mit MQTT und Node-Red. Beides ist für mich völliges Neuland, möchte aber sehr gerne in diese Materie einsteigen. Beides ist inzwischen (unter Windows) schon installiert. Aber ich bekomme es zum Verrecken nicht hin, Daten von Open TDU zu sehen... Nun habe ich entdeckt, dass unter Einstellungen - Info - MQTT steht, dass MQTT getrennt ist. Was bedeutet das???? und warum könnte es so sein???. Mir ist klar, dass es mit Sicherheit nicht eine einzige Antwort geben kann, sondern mit Sicherheit tausende Möglichkeiten geben kann. Aber vielleicht gibt es ja doch ein Hinweis mit dem ich was anfangen kann. MQTT ist natürlich unter Einstellungen aktiviert.

tschau
Wolfgang

**Comment by helgeerbe** [comment]
Hallo @wolwes, MQTT ist ein Broker Protokoll. D. h. du brauchst einen Service wie z. B. mosquitto. Dieser Service nimmt Nachrichten entgegen und verteilt sie.

OpenDTU z. B. veröffentlicht (publish) Nachrichten unter sogenannten topics. Mit Node Red könntest du jetzt auf diesen Topics lauschen (subscribe). 

Es ist aber keine bilaterale Schnittstelle sondern 1:N und darüber hinaus auch mit unterschiedlichen Qualitätsstufen, um den garantierten Empfang von Nachrichten sicherzustellen (Stichworte Quality of Server, Retain Flag).

Einfach mal mosquitto tutorial googeln.

**Reply by wolwes** [reply]
Hi,
danke für die Tipps, aber das ist mir alles bekannt. MQTT läuft bei mir, durch..._sub und ..._pub getestet. Ich verfolge im Moment der Idee, dass das was mit den in Open DTU **nicht** vergebenen Benutzer und Passwort zu tun hat. Es kann sein, dass ich die conf-Datei noch mit anonymus=true ergänzen muss. 
tschau

**Reply by ThomasManMan** [reply]
Benutze den http://mqtt-explorer.com/ . Damit schaust du dir die Daten an,  die am mosquitto eintrudeln.

**Comment by DPO99** [comment]
Welchen MQTT Broker verwendest du denn?

**Comment by wolwes** [comment]
mosquitto

**Reply by wolwes** [reply]
Den mqtt-explorer kannte ich noch nicht.... Danke für den Tipp. Ich kann in diesem Explorer sehen, dass mosquitto funktioniert. Aber warum kriege ich keine Daten von Open DTU??? Topic habe ich auf solar/ gelassen. Wenn ich im Explorer z.B. solar/# eingebe, kommt nichts...

**Comment by DPO99** [comment]
Also wenn mqtt nicht verbunden ist dann wird das so sein. Gewisse werte wie die Strings und die Ertragsdaten sollten m.e. automatisch an den mqtt Broker übergeben werden. Anmeldedaten für mqtt sind aber korrekt? Falls leer ggf. Mal befüllen, ich habe das noch nie ohne Anmeldendaten versucht.

**Comment by wolwes** [comment]
ich glaube auch anonym geht wohl nicht. Werde gleich mal Benutzer und Passwort eingeben. Dann schauen wir mal... Ich werde aber wohl erst gegen 22.30 Uhr dazu kommen....

**Comment by DPO99** [comment]
Ich glaube es geht schon ohne aber das musst du dann in der config datei des Brokers erst einstellen. Mit User und Passwort kriegst ja die Verbindung hoffentlich erstmal hin und kannst dann weiter rum spielen.

**Comment by wolwes** [comment]
OK, ich werde das mit Benutzer und Passwort jetzt so machen. Aber die Wirkung werde ich momentan nicht sehen, weil die, der das DTU nachts nichts Vernünftiges sendet....Also bis demnächst....

**Reply by wolwes** [reply]
Benutzername+Passwort vergeben, scheint aber nichts zu nützen...
Deshalb jetzt ein Screenshot. Vielleicht kann sich der Eine oder Anderer einen Reim darauf machen. Wäre schön, wenn das mal gelöst werden könnte. Beschäftige mich schon seit Tagen intensiv damit und komme trotzdem zu keinem Ergebnis.
![MQTT](https://user-images.githubusercontent.com/16593975/228223332-6c877793-3729-422a-8f4a-96cfd2e63c29.png)

**Comment by tbnobody** [comment]
Mach doch mal einen screenshot von deiner mqtt einstellungsseite

**Comment by wolwes** [comment]
Benutzer und Passwort habe ich wieder rausgenommen, war ja nicht "schuldig".
![screen 1](https://user-images.githubusercontent.com/16593975/228239891-a2a7b72e-6695-4156-95b3-8309e85200fe.png)
![screen 2](https://user-images.githubusercontent.com/16593975/228239959-b9baf19a-5cdc-426a-a081-f3f29a55f9d8.png)

**Comment by tbnobody** [comment]
127.0.0.1 ist als IP mal definitiv falsch. Die IP verweist ja auch localhost, Also OpenDTU selbst und nicht auf den MQTT Broker.

**Reply by wolwes** [reply]
ok, und was und wie bekomme ich eine richtige IP?. Ich habe die wohl irgenwo her, also nicht einfach so eingetragen... OK, Halbwissen oder gar kein Wissen meinerseits.

**Comment by DPO99** [comment]
Das ganze HomeAssistant Zeug kannst du m.E. auch ausschalten...du nutzt ja einen anderen Broker (aber das hat nichts mit Deinem Problem zu tun).

**Reply by tbnobody** [reply]
Home Assistant hat auch nichts mit dem Broker zu tun. Es dient der Auto Discovery Funktion für Home Assistant. Aber wenn keine Home Assistant Software verwendet wird, kann das in der Tat ausgeschaltet werden. Wie gesagt, die Broker IP ist falsch.

**Comment by wolwes** [comment]
Klar, HomeAssistant habe ich aktiviert, da ich vorhabe, mich damit zu beschäftigen. Habe ich auch schon auf dem rasp 4 installiert. Bin aber noch nicht sehr weit damit gekommen.

**Reply by tbnobody** [reply]
Ggf. solltest du aber den Broker usw. auch auf dem Pi installieren? Ansonsten musst du ja deinen Windows PC immer anlassen.

**Comment by wolwes** [comment]
jetzt weiß ich auch wieder, woher ich die IP habe: Aus der Make -Node-Red Spezial, Seite 16. Dort wird beschrieben, was und wie Mosquitto installiert wird...

**Reply by tbnobody** [reply]
Du trägst ins NodeRed aber die localhost IP ein weil der Mosquitto am gleichen Host läuft wie NodeRed. 
Welche IP hast du denn in MQTT-Explorer eingetragen?

**Reply by wolwes** [reply]
localhost

**Reply by tbnobody** [reply]
wo läuft dein broker und wo hast du mqtt explorer installiert?

**Reply by wolwes** [reply]
auf dem Rechner, auf dem wir hier schreiben, also unter Windows..

**Reply by tbnobody** [reply]
Dann trage doch in OpenDTU die IP von deinem Rechner ein (Wenn du eine Fritzbox hast ist die IP ziemlich sicher sowas wie 192.168.178.xxx). Ggf. musst du aber noch den MQTT Port in der Windows Firewall freischalten damit der MQTT Datenverkehr nicht beblockt wird.

**Reply by wolwes** [reply]
ok, die Marschrichtung ist jetzt klar. Werde mich jetzt darum kümmern. Melde mich dann wieder. Dauert aber wohl noch ein wenig...

**Comment by wolwes** [comment]
an einer Stelle in der Firewall habe ich unsicher gestutzt: UDP oder TCP als Protokoll??? Ich habe TCP genommen, weil das ja auch im Logbuch so erwähnt wurde. Ansonsten habe ich nun die IP von meinem PC eingetragen und eine neue Regel in der Firewall erstellt.

**Reply by wolwes** [reply]
Wow!!!
es hat geklappt... Die Disconnect -Meldungen im Logbuch sind weg. Im MQTT-Explorer werden solar-Werte angezeigt...Das passierte aber erst, nachdem ich Windows neu gebootet habe. Das Ganze auch als Anonymer-Benutzer. Ich habe dazu aber auch 2 Einträge in conf-Datei vom Mosquitto vorgenommen: anonymus=true und irgendwas mit listener...

Das war ganz toll, was Du - tbnobopdy - hier geleistet hast. Ich danke Dir, dass Du mich nicht aufgegeben und weiterhin geholfen hast. Ich habe jetzt ne Menge gelernt und bin sehr glücklich darüber, dass ich mich hier gemeldet habe.

Ansonsten ist mir klar geworden, dass ich unbedingt den Broker auf dem Raspberry 4 installieren sollte. 
tschau
Wolfgang

**Reply by wolwes** [reply]
und wie kriegen wir jetzt das unanswered weg???

Alles klar. Ich bin begeistert. Alle Fragen wurden beantwortet....
